text/1004-transactions-api.rst

```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
+.. _Execute: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-execute
+.. _ExecuteScript: https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-execute-script
+
+
+Future Work
+===========
 
-To make changing EdgeDB address or restarting EdgeDB work nicely, we
-need to retry simple queries on connection errors too:
+
+More Configuration
+------------------
+
+Setting tracing/debugging metadata for queries may be implemented
+in the same way:
 
 .. code-block:: python
 
-    pool.query("SELECT ..")
+    def handle(request, db):
+        db = db.with_metadata(
+            uri=request.uri,
+            username=request.session.get('username'),
+        )
+        handle_user_request(request, db)
+
+And this can be done on transaction level too:
+
+.. code-block:: python
+
+    async for tx in db.retry():
+      async with tx:
+        tx = tx.with_metadata(
+            uri=request.uri,
+            username=request.session.get('username'),
+        )
+        tx.query("...")
+
 
-But there are couple of issues:
+More Read-Only Options
+----------------------
 
-1. Repeating read-only queries is always safe, but we don't know which
-   ones are readonly. We tackle this below.
-2. Repeating non-readonly queries can be dangerous: they may be applied
-   twice.
+Replica config may be specified when configuring a read-only connection:
 
-There are couple of ideas to differentiate read-only and mutable
-queries:
+.. code-block:: python
 
-1. We can expose "read-only" flag in ``Prepare`` (and it's always safe
-   to retry before ``Execute`` happens)
-2. Enable connection mode that forbids mutating queries outside of
-   transaction and retry all (read-only) queries
-3. Do (2) and have some method to override it on per-method-call basis
-4. Or vice versa add ``pool.read_only_query()`` method
-5. Add ``pool.read_only().query()``
-6. Add ``pool.with_options({readonly: true}).query()``
+   def read_only(self, primary=False, max_replica_lag=10): ...
 
-Note: methods (2-6) are also helpful for working with primary/replica
-installations. But probably only last two would allow full power, as they allow
-``pool.read_only(primary=true)`` (i.e. in case you need read-only transaction
-that can't go to a replica).
+Perhaps this should be encapsulated into replica options:
 
-This issue can be solved by a later RFC.
+.. code-block:: python
+
+   def read_only(self, primary=False, replicas: ReplicaOptions): ...
 
 
 Learning Curve
 ==============
 
-This complicates learning curve, but:
+The ``retry`` method complicates learning curve curve, but:
 
 1. This is already a problem in Postgres, there a lot of people who
-   ignore the issue for pet projects and a lot of startups having "normal"
-   rate or 500 errors at any point of time, while it's preventable.
+   ignore the issue for pet projects and a lot of startups having
+   "normal" rate or 500 errors at any point of time, while it's
+   preventable.
```

Comment:
"1. Letting the application to error out instead of automatically retrying certain transactions is a wildly known, yet an entirely preventable problem."